### ISSUE TITLE
Kea 1.0

### ISSUE THREAD

**Issue by mariusandra** [keajs/kea#98]
Hi all, 

I've been [rewriting Kea](https://github.com/keajs/kea/tree/big-rewrite) in order to reach the big `1.0` milestone for several days now. I'm nearly there. However before releasing anything, I'd like to ask the community for help with testing and ironing out any edge cases that might show up. 

So please, if you're using kea in your app, read the "How to test?" section below, try it out and report any issues you may experience. 

### Complete rewrite of the internals

A near complete rewrite of the internals, resulting in a bundle size decrease from `16kb` to `10kb` (minified, not compressed): ![image](https://user-images.githubusercontent.com/53387/56729772-5a385380-6756-11e9-98ba-fce6825135b2.png)

Code complexity went down a lot, leading to an improved plugin experience (more on that later).

### Much much simpler dynamic logic stores

Previously if you used many dynamic logic stores, you probably ran into cases where you had to check `payload.key === key` everywhere. For example:

```js
kea({
  key: (props) => props.id,
  path: (key) => ['scenes', 'counterDynamic', 'counter', key],

  actions: () => ({
    increment: (amount = 1) => ({ amount }),
    decrement: (amount = 1) => ({ amount })
  }),

  reducers: ({ actions, key, props }) => ({
    counter: [0, PropTypes.number, {
      // don't do this anymore
      [actions.increment]: (state, payload) => payload.key === key ? state + payload.amount : state,
      [actions.decrement]: (state, payload) => payload.key === key ? state - payload.amount : state
    }]
  }),
})
```

Or worse, in saga workers:

```js
kea({
  workers: {
    doSomething: function * (action) {
      if (action.payload.key !== key) { // don't do this anymore
        return
      } 
      // do stuff
    }
  }
})
```

This is no longer necessary. 

Now all actions created by every individual dynamic logic store are unique and **`action.payload.key` doesn't exist anymore**.

This does mean that you can't share actions between dynamic logic stores like you could before. If you still need actions that are common to all instances of one dynamic logic store, create them in a separate `kea({})` call and connect to them from your dynamic logic store.

### Removed `selectors.root`

If you are using `logic.selectors.root` anywhere, switch to `logic.selector`.

Similarly, change code like this:

```js
kea({
  connect: { props: [logic, 'root'] }
})
```
to this:

```js
kea({
  connect: { props: [logic, '* as root'] }
})
```

### Track mounting and un-mounting of logic stores internally

Before all plugins needed to do black magic to find out which logic stores are currently in use, which are actively mounted on the page, which just got unmounted, etc. Now tracking this is integrated inside kea and to for example start or stop sagas when a component mounts, you just integrate the plugin hooks `mounted(path, logic)` and `unmounted(path, logic)`. 

### Upgraded plugin infrastructure:

The refactored code makes 

The interface for plugin makes is simplified and very straightforward. Just create an object with the following structure and drop it into a `plugins: [myPlugin]` array. Either on the store to activate globally or in `kea({ plugins: [myPlugin] })` to activate just for one logic store.

Plugins need to have the following structure, where `logic` is the logic store actively being created and `input` is the input in `kea(input)`.

```js
    const plugin = {
      // Required: name of the plugin
      name: 'myPlugin',

      // Run before the redux store creation begins. Use it to add options (middleware, etc) to the store creator.
      beforeReduxStore (options) {},

      // Run after the redux store is created.
      afterReduxStore (options, store) {},

      // Run before we start converting the input into the logic
      beforeCreate (logic, input) {},

      // Run after each step in the conversion
      afterConnect (logic, input, addConncetion) {},
      afterConstants (logic, input) {},
      afterActions (logic, input) {},
      afterReducerInputs (logic, input) {},
      afterReducers (logic, input) {},
      afterReducerSelectors (logic, input) {},
      afterSelectors (logic, input) {},

      // Run after the input is fully converted to the logic
      afterCreate (logic, input) {},

      // Run when a logic store is mounted/unmounted in React
      mounted (pathString, logic) {},
      unmounted (pathString, logic) {},

      // Run after mounting and before rendering the component in React's scope (you can use hooks here)
      beforeRender (logic, props) {},

      // Run when we are removing kea from the system, e.g. when cleaning up after tests
      clearCache () {}
    },
    ...
```

### New plugins coming:

These are in incubation for now and will be released soon.

#### kea-listeners

Similar to `takeEvery` in `kea-saga`, listeners listen for actions and let you run and dispatch code after they run:

```js
  kea({
    listeners: ({ actions }) => ({ 
      [actions.updateName]: (action, store) => {
        console.log(action.payload)
        store.dispatch(actions.updateDescription('asd'))
      },
    })
  })
```

I'm doubting if this should be a plugin or even part of the core kea package. Having it part of the core opens up a lot of extensibility options for future plugin authors.

#### kea-router

Built with `kea-listeners`, `kea-router` acts as a bridge between `kea`, `react-redux` and `connected-react-router`. 

```js
  kea({
    actionToUrl: ({ actions }) => ({
      [actions.selectEmail]: () => `/signup?type=email`,
      [actions.unselectEmail]: () => `/signup`,
      [actions.openLesson]: ({ id }, action) => `/open/${id}`
    }),

    urlToAction: ({ actions }) => ({
      '/signup?type=email': url => actions.selectEmail(),
      '/signup': url => actions.unselectEmail(),
      '/open/': (url) => actions.openLesson(parseInt(url.split('/')[2]))
    }),
  })
```

It's pretty crude at this point. Proper pattern matching will come later.

### Old plugins upgraded

`kea-saga`, `kea-thunk` and `kea-localstorage` have been upgraded to work with kea 1.0

### Lazy mounting mode

Now you can choose whether to have all `kea()` calls add themselves to the redux store when first possible... or if it should be done lazily, when the component is mounted.

Lazily created logic stores also unmount from redux when the component unmounts.

The API for this is still experimental (`kea({ options: { lazy: true })`) and can change. 

By default static logic stores (no key) are loaded eagerly, dynamic logic stores (with a key) are loaded lazily. Same for static logic stores that connect dynamically (`connect: { props: [otherLogic.withKey(props = props.id), ['something']] }`). 

There's still a bit work to do here, e.g. tracking if eager logic stores want to connect to lazy logic stores and making them lazy as a result.

### Possibility to initialize before the store is created

Still to do: now with lazy logic stores, we can theoretically have `kea()` calls work before the store is initialized. They will just be connected to the store when mounting the components.

I haven't tested it yet, but it should be possible. This will get rid of some edge cases with 

### Way to merge new logic into old logic

There is now an experimental `kea({ ..., merge: [{ reducers: ({}) => ({}) }] })` merge function that combines whatever logic it is given in the `input` format from `kea(input)`. 

### Many edge case bugs are solved

More on them later. 

## How to test?

Make sure you're running `react-redux` version 7.0 or later and `react` version `16.8.3` or later.

Upgrade all your packages to the latest beta versions:

- kea: ![kea](https://img.shields.io/npm/v/kea/beta.svg)
- kea-saga: ![kea](https://img.shields.io/npm/v/kea-saga/beta.svg)
- kea-thunk: ![kea](https://img.shields.io/npm/v/kea-thunk/beta.svg)
- kea-localstorage: ![kea](https://img.shields.io/npm/v/kea-localstorage/beta.svg)

... and then please report if it works fine, what broke, did you notice any performance or other issues, etc.

I've been running a large production webapp on the latest 1.0 beta now for a while and so far so good. Other than the changes that are documented above (mostly removing `payload.key` checks), everything went smooth with the upgrade.

**Comment by mariusandra** [comment]
Hi, I'm tagging a few people who might be interested in this and could possibly test and provide feedback. Please just ignore/unwatch if you're not using kea anymore or find this irrelevant for your case: @meikoudras @madisvain @xantorres @patrick-jones @amrdraz @lovekaizen @nikolakanacki @ingoclaro @neginbasiri @ingoclaro @guyzmo @tonyhb @prsauer @k1sul1 

Thank you! 😊

**Comment by amrdraz** [comment]
I can't test as I'm occupied with other projects. for all it's worth I moved on from using react to vue with my recent work and therefor kea, but I still see kea as a better approach to building redux based state stores and I'm happy how the project progressed.

**Comment by mariusandra** [comment]
There might be a memory leak with `1.0` if you're doing SSR since we now cache the logic that is created on the page. During a long running SSR session this might mean all the logic ever made will be cached.

One solution seems to be running `resetKeaCache()` after `renderToString()`

```js
import { resetKeaCache } from 'kea'
const html = renderToString(<...>)
resetKeaCache()
```

I need to see if there is a better way to handle this...

**Comment by etiennemtl** [comment]
There seems to have a problem with how the connect function wraps a Component. I've been trying to call saga actions using next.js, `static async getInitialProps` lifecycle method, but it's undefined in the WrappedComponent provided by the connect function.

**Comment by mariusandra** [comment]
Hey!

I played around with it and got the `getInitialProps` working with kea `1.0.0-beta.17`.

The issue comes from the fact that when you wrap any Page with kea, it wraps it with its own `<Kea/>` component... which of course does not implement `getInitialProps`.

So the solution is to manually define the `getInitialProps` method on the `<Kea />` component... and tell it to call the one of the page.

The easiest way to do this is to create a plugin that injects this for us. Here is an example of such a plugin:

```js
// in store.js
import { getStore, activatePlugin } from 'kea'

activatePlugin({
  name: 'next',
  afterWrapper (input, Klass, Kea) {
    if (Klass.getInitialProps) {
      Kea.getInitialProps = async function (ctx) {
        await Klass.getInitialProps(ctx)
      }
    }
  }
})

export const initStore = () => getStore({})
```

The `afterWrapper` hook was not present before `beta.17`, so it will not work before.

This is also just a work in progress... and I already found a small issue. For example on a [next.js and kea test app](https://github.com/keajs/kea-next-test), I [call](https://github.com/keajs/kea-next-test/blob/6d8af0c7d84ae72a69de020263ae91c44e51964c/pages/index.js#L48) the `increment` action in the `Index` page. It runs on the server, yet when the client loads it, the `counter` is still at `0`. If I click on the "home" link, it stays at 0.

Yet if I open `/about` as the first page, then every time I click on "Home" the counter increases by one.

You may try it live here: https://nextjs.mariusandra.now.sh/

Eventually this should be released as a `kea-next` package to offer first class support for next.js projects :).

I must run now, I'll check this in more detail and reply in your other issue about sagas the first chance I get.

**Comment by k1sul1** [comment]
I'm about to start a new rewrite of a side project of mine, I'll definitely give it a go there. Looks nice.

I might have a work project soon where we're going to use React, if Redux is warranted there, I'll be sure to propose using Kea for that as well.

**Comment by etiennemtl** [comment]
@mariusandra Thanks a lot for taking the time to have a look at this, if I can help in any way. I will try to start a repo for a kea-next plugin. It would be necessary also to support other plugins like kea-saga, kea-listeners etc.. I think it would be a great addition to the kea library to be able to fully support nextjs.

**Comment by mariusandra** [comment]
@etiennemtl thanks for bringing this issue to the forefront!

I played around with next.js a bit today and found three issues that need to be resolved cleanly:

1. How to activate plugins so that they are active for all `kea({})` calls. The issue is that the `initStore` function in `store.js` is called after the `kea({})` functions in the `*.js` files in `/pages`. So the plugins might not be active by the time `kea({})` is called. Running `activatePlugin()` directly in the [global scope](https://github.com/keajs/kea-next-test/blob/master/store.js#L3) of `store.js` seems to mostly work, but it's not exactly reliable.

2. The SSR issue mentioned a few comments above still needs to be resolved. Right now if you have logic stores defined without a path (defaulting to `kea.inline.${++globalCounter}`), then every server render increases globalCounter and the next logic has a different path than before. Plus they don't get cleared after rendering and stay around in the state, probably even if a different user reloads the page. I have ideas to solve it, but had no time to implement them yet. The biggest issue here is that without reliable paths for the logic stores, we can't hydrate the store from SSR on the client side. 

3. All this could be solved with lazy logic. The here issue is then kea relies on React to mount the components and connect them to the redux store. However next.js's `getInitialProps` runs before React renders the page and thus there are no actions to call and nothing in the store to select. 

One idea is to develop a system that could mount the logic on redux without requiring react to be present. So something like this:

```js
Index.getInitialProps = async function (ctx) {
  indexLogic.mount().then(({ actions, selectors }) => {
    ctx.store.dispatch(actions.increment(1))
    const counter = selectors.counter(ctx.store.getState())

    // or perhaps we could bind dispatch and getState automatically and run it like this:
    actions.increment(1)
    const counter = selectors.counter()
  })
}
```

So still some things to test. I'll be pushing my progress to the repo here: https://github.com/keajs/kea-next-test

**Comment by etiennecl** [comment]
I've been playing around with the next example, I've been trying a lot of different things regarding mounting the logic manually inside getInitialProps, although without success. Could you point me in the right direction?

**Comment by tonyhb** [comment]
We've moved over to typescript/hooks and are using `easy-peasy` for the time being.  I much prefer the separation of actions, reducers, thunks etc. that Kea provides (rather than a flat object), but I'll see if I can give it a test some time soon.

**Comment by mariusandra** [comment]
Hi @etiennecl , thank you for the continued work on this. I indeed made some big changes regarding context in the latest beta (v20). Previously kea had an "internal context" called a cache, which was set when your bundle loaded and reset when forced to via `resetKeaCache`. This worked fine in the browser, when the cache was always empty on each page load and never reset, but it caused issues with SSR and frameworks like next.js.

With the changes in beta.20, this cache is now explicitly called a context and there are tools to mainpulate it. (Functions like `getContext`, `setContext`, `resetContext`, `withContext`, etc). I will write better documentation about it as soon as I can.

However until then, you should run `resetContext()` before any call to `kea` is made if you wish to configure defaults, plugins, etc. You should no longer pass plugins to the store. Instead pass them to the context.

Another big change is that previously we would store all the data inside logic stores (`const logic = kea({})`). Now logic stores are just a collection of functions that delegate to the context. So theoretically if you swap out the context, everything else should still work.

Yet another big change is that the lifecycle of kea logic is also more clearly defined. A logic can be in any of these 3 states: 1) initialized, 2) built, 3) mounted.

1. **initialized:** Now when you call `kea({})` the logic store is just initialized and added to the context (to `context.inputs`). Nothing will be automatically built not mounted.

2. **built**: This refers to converting the kea input (`kea(input)`) into a logic (object with keys `actions`, `reducers`, `selectors`, etc). Previously this happened automatically in step 1, but now it's deferred as late as possible.

3. **mounted:**. A logic is mounted if it is connected to redux. This is also when all the sagas and other side effects run and do what they need to do. This happens automatically when you wrap a logic around a React component. If a component is not built when you try to mount it, it will be built automatically.

Before 1.0 all kea logics (at least the ones without a `key`) were automatically mounted when they were initialized. This is no longer the case. If you wish to revert to the pre-1.0 behaviour, run `resetContext({ autoMount: true })` before any calls to `kea()` are made. 

(NB! `kea({ options:{lazy: true} })` no longer works. It's all configured from the context now. I need to update the docs...)

Like I said, logics are automatically mounted when you wrap them around react components. However if you need to mount them manually, it's also possible:

```js
const logic = kea({ ... })

resetContext({ ... })

const unmount = logic.mount()

const { store } = getContext()

store.dispatch(logic.actions.bla())
const value = logic.selectors.getStuff(store.getState())

unmount()
```

Mounting like this should also start the sagas.

For next.js and the getIinitialProps, I still need to experiment with it and see how it could all work. The building blocks are all there, just I need to make sure they fit well. I'll get to it in the next days. It's possible the next redux wrapper needs to be opened up and adjusted to make it all work.

I'll share more news when I have any. Cheers!

**Comment by k1sul1** [comment]
I've worked with this for a while now, everything seems to be running smoothly. Can't wait for the new plugins (all of them), because I've been relying on black magic like this.

```
export const actionResultIs = (action, value) => {
  const name = action.toString()
  return ({ type, payload }) => {
    if (name === type && Array.isArray(value)) {
      return value.some(x => x === payload)
    }

    return name === type && payload === value
  }
}

export const actionResultObjectContainsKeys = (action, keys = []) => {
  const name = action.toString()

  return ({ type, payload }) => {
    return name === type && keys.some(key => key in payload)
  }
}

[actions.initialize]: function * () {
  yield put(actions.setStatus('loading'))

  yield put(actions.getPrimaryMenu())
  yield put(actions.loadPage(`${window.location.pathname}${window.location.search}`))

  // wait for the page to load
  yield take(actionResultObjectContainsKeys(actions.setPageData, ['status', 'post_type', 'id']))
  yield put(actions.setStatus('ready')) // The menu has a loading indicator, app can be started without waiting for it
},
```

Listeners seem like a much better choice.

**Comment by mariusandra** [comment]
Hi everyone,

As of `beta.31` we now support basic hooks!

```js
const logic = kea({  actions: () => ({
    updateName: name => ({ name })
  }),
  reducers: ({ actions }) => ({
    name: ['Bob', {
      [actions.updateName]: (_, payload) => payload.name
    }]
  })
})

function NameComponent (props) {
  const { name } = useProps(logic)
  const { updateName } = useActions(logic)

  return (
    <div>
      <div>Name: {name}</div>
      <button onClick={() => updateName('George')}>Change</button>
    </div>
  )
}
```

It's still a work in progress and there are some gotchas, but it's getting there.

I put up a long doc with all the 1.0 changes here:
https://github.com/keajs/kea/blob/master/docs/CHANGES-1.0.md

**Comment by mariusandra** [comment]
After 2+ months of work over 260 commits, `1.0.0-rc.1` is finally here. I expect little to change before `1.0.0` final is out. Now it's just time to work on updating all the documentation, articles, etc.

Please try it out and share your feedback here.

I have also extensively updated the [1.0 changes document](https://github.com/keajs/kea/blob/master/docs/CHANGES-1.0.md). 

(... and feel fee to RT [this tweet](https://twitter.com/mariusandra/status/1141102464772595712) as well...).
